app/models/watching_transaction.rb

```diff
@@ -0,0 +1,79 @@
+# frozen_string_literal: true
+
+require 'ethereum_api'
+
+class WatchingTransaction < ApplicationRecord
+  before_create :set_uuid
+
+  belongs_to :user
+  has_many :transactions, class_name: 'WatchingTransaction', foreign_key: 'group_id'
+  belongs_to :original, class_name: 'WatchingTransaction', foreign_key: 'group_id', optional: true
+
+  validates :transaction_object,
+            presence: true
+  validates :signed_transaction,
+            presence: true
+  validates :txhash,
+            presence: true,
+            uniqueness: true
+
+  def set_uuid
+    self.id = SecureRandom.uuid
```

Comment:
You mean instead of...
self.group_id = id unless group_id
Do...
self.group_id |= id
?
And why test should be able to override id? Isn’t id autogenerated all the time?

Reply:
The second one is correct `self.group_id |= id` which implies if the `group_id` is `nil`, it uses the `id` itself.

For the second point, it is so that you can use `FactoryBot` to create group of transactions together. For example, when testing the `resend_transaction`, you can create a group of transactions like so:

```ruby
original_transaction = FactoryBot.create(:watching_transaction)
4.times do 
  FactoryBot.create(:watching_transaction, group_id: original_transaction.group_id)
end

separate_transaction = FactoryBot.create(:watching_transaction)

WatchingTransaction.resend_transactions()
```

For the test setup to work, FactoryBot should be able set the `group_id`. If  we have `self.group_id = id`, it will create different transactions itself.

Reply:
My question was about `id` and not about `group_id`
Why do you want non-null assignment for `id` also?

Reply:
Anyways it's not working: NoMethodError: undefined method `|' for "1":String
Why `self.group_id = id unless group_id` is not good?
And why tests should be able to override `id`?

Reply:
For `group_id` and `id` have the same reason but more strongly with `group_id`. Tests don't necessarily have to override it but for auto-generated fields should have an option to be manually assigned.

One example for local development testing is to have fixed IDs for seeding. It is easier to type `User.find("1")` than `User.find("abcd-efgh")`.  Of course, `id` cannot be overridden by the API interface. But internally it is a good feature or default to allow them to be manually set in console.

Reply:
To make it work, you need `self.id ||= SecureRandom.uuid`. 

I keep making a mistake on the `|` or `||` for default assignment.